### ISSUE TITLE
Resnet wrong results with Nvidia/caffe

### ISSUE THREAD

**Issue by Wronskia** [NVIDIA/caffe#433]
Hello,

I am trying to reproduce the results of https://github.com/cvjena/cnn-models/tree/master/ResNet_preact/ResNet50_cvgj for resnet50 architecture. However I get poor results on inference  (i.e 0%) while getting the stated results in caffe (Around 74% for the top1 Acc). 

do you have an idea from where the problem could arise? 

Thank you,
Best

**Comment by drnikolaev** [comment]
Hi @Wronskia 
to answer this question we need at least to see what you do exactly.

**Comment by Wronskia** [comment]
Hello @drnikolaev,
Sorry for the lack of details.

Here is what I did : 
I used the test.prototxt of : https://github.com/cvjena/cnn-models/tree/master/ResNet_preact/ResNet50_cvgj  with the pretrain weights that they provide which are available here : https://drive.google.com/open?id=0B6VgjAr4t_oTTDh2SVJIa2VkZVU


The test.prototxt I am using :  [test.prototxt.docx](https://github.com/NVIDIA/caffe/files/1393898/test.prototxt.docx) 

On top of that, I subsampled Imagenet dataset to around 10000 images which I rescaled keeping the aspect ratio. 
I then run the inference using this : 

../../../opt/caffe/caffe/build/tools/caffe test \
--model="test.prototxt" \
--iterations="1000" \
--weights="resnet50_cvgj_iter_320000.caffemodel" \
--gpu "0" 2>&1 | tee run.log
~                          

and I get the following results :  

I1017 17:11:08.610745   462 caffe.cpp:303] Batch 999, accuracy = 0
I1017 17:11:08.610770   462 caffe.cpp:303] Batch 999, loss = 74.4608
I1017 17:11:08.610780   462 caffe.cpp:303] Batch 999, top5 = 0
I1017 17:11:08.610787   462 caffe.cpp:308] Loss: 80.4354
I1017 17:11:08.610867   462 caffe.cpp:320] accuracy = 0.0014
I1017 17:11:08.610913   462 caffe.cpp:320] loss = 80.4354 (* 1 = 80.4354 loss)
I1017 17:11:08.610924   462 caffe.cpp:320] top5 = 0.0054

while doing the exact same thing in caffe, I get around 74% in top1 Acc.

Please find attached the test.prototxt I am using

Thank you very much,
Best.

   
[test.prototxt.docx](https://github.com/NVIDIA/caffe/files/1393898/test.prototxt.docx)

**Comment by Wronskia** [comment]
Hello @drnikolaev ,

I performed inference for a single image using deploy.prototxt given here :  https://github.com/cvjena/cnn-models/tree/master/ResNet_preact/ResNet50_cvgj along with pycaffe.
I then extracted the numpy arrays of each layer after inference on that image and calculated the L2 norm to see the difference between caffe and Nvidia/caffe.



| Layers   |      L2 norm Nvidia/caffe       |  L2 norm caffe |
|----------|:-------------:|------:|
| Layer 0 data blob  |  24986.2  | 24986.2|
| Layer 1 batchNorm Layer |    19470.7  |   561.849 |

The difference between both batchnorm layers is quite large. for this specific image, caffe classifies correctly while Nvidia/caffe doesn't.

Please let me know if there is something I am missing,

Thank you very much,
Best regards.

**Comment by Wronskia** [comment]
Hello @drnikolaev ,

I put the files which I used to get these results with a small description in this repo https://github.com/Wronskia/inference-results-NVcaffe

would it be possible to provide the .caffemodel of your resnet50 that is compatible with NVcaffe so that I can use it as pretrain model ?

Thank you very much,
Best

**Comment by drnikolaev** [comment]
hi @Wronskia we can't distribute models due to copyright concerns

**Comment by drnikolaev** [comment]
@Wronskia thank you for reporting this. We have reproduced the issue and now working on the fix

**Comment by whria78** [comment]
I had a similar problem. 

When I tried to fine-tuning ResNet-152, 
it always showed 0 accuracy. It failed to finetune.

ResNet-152 pretrained model =from  https://github.com/KaimingHe/deep-residual-networks
 
solver.prototxt and trainval.prototxt =
[resnet-152.zip](https://github.com/NVIDIA/caffe/files/1424803/resnet-152.zip)


nvcaffe - ResNet-152 finetuning -> Fail
BVLC - ResNet-152 finetuning -> OK
nvcaffe - VGG 19 -> OK
BVLC - VGG 19 -> OK

I tested in CUDA 8.0 + cuDNN 7.0 and CUDA 7.5 , cuDNN5.1.

**Comment by drnikolaev** [comment]
hi @Wronskia I'd like to clarify what images were used to train. All I see is this:
```
 root_folder: "/home/atlas1_ssd/simon/ilsvrc12-scaled/train/"
```
So, _scaled_ how? Keeping aspect ratio? To square? When I train&test RN50 I see correct results. I suspect that in your case something don't match. Please clarify.

**Comment by Wronskia** [comment]
hello @drnikolaev 

the scripts I use for rescaling are available here : https://github.com/Wronskia/inference-results-NVcaffe (I rescale while keeping the aspect ratio). I also have reasonable results while finetuning RN50 with imagenet however when I just perform inference without training I get 0%.

with the test.prototxt I provide on top
../../../opt/caffe/caffe/build/tools/caffe test 
--model="test.prototxt" 
--iterations="1000" 
--weights="resnet50_cvgj_iter_320000.caffemodel" 
--gpu "0" 2>&1 | tee run.log
~

Thanks,
Best

**Comment by drnikolaev** [comment]
Please check v0.16.5 and reopen the issue if the problem still exists.

**Comment by whria78** [comment]
Thank you for releasing v0.16.5.

I had an error (.... libturbojpeg.a: error adding symbols: Bad value ....) when compiling v0.16.5, but I could solve it by

sudo ln -s /usr/lib/x86_64-linux-gnu/libturbojpeg.so.0.0.0 /usr/lib/x86_64-linux-gnu/libturbojpeg.so

Anyway, I am now in fine-tuning the ResNet-152. 

Until now, it works well. It showed a normal Test net output and Accuracy.

Thank you again.
